src/case_encoder.h

```diff
@@ -72,34 +75,53 @@ class UpperCaseEncoder : public CaseEncoder {
   UpperCaseEncoder(bool removeExtraWhiteSpace)
   : removeExtraWhiteSpace_(removeExtraWhiteSpace) {}
 
-  std::pair<absl::string_view, int> normalizePrefix(absl::string_view input) {
+  std::pair<absl::string_view, int> normalizePrefix(absl::string_view orig_input) {
+    if((dump_buffer_from_ >= 0) && (dump_buffer_from_ < buffer_queue_.size())) {
+      return buffer_queue_[dump_buffer_from_++];
+    }
+
+    if(dump_buffer_from_ > -1) {
+      dump_buffer_from_ = -1;
+      buffer_queue_.clear();
+      return {nullptr, 0};
+    }
+
+    const auto input = orig_input.substr(offset);
     auto p = CaseEncoder::normalizePrefix(input);
     auto sp = p.first;
     int consumed = p.second;
 
     bool last = input.size() == (size_t)consumed;
     decltype(p) ret;
 
-    auto null = [](int consumed) -> std::pair<absl::string_view, int> {
-      return {{nullptr, 0}, consumed};
+    auto null = [this](int consumed) -> std::pair<absl::string_view, int> {
+      offset += consumed;
+      return {{nullptr, 0}, 0};
     };
 
-    auto buffer = [this](absl::string_view sp) {
+    auto buffer = [this, p](absl::string_view sp, int override_consumed = -1) {
+      auto cur_buf_last = buffer_.size();
       buffer_.append(sp.data(), sp.size());
+      auto tmp_str = std::string(buffer_).substr(cur_buf_last, sp.size());
+      buffer_queue_.push_back({tmp_str, override_consumed == -1 ? p.second : override_consumed});
```

Comment:
As above, consider adding a few comments here, so that it will be easier to recall what is being done here.